src/lightgbm/src/main/scala/LightGBMUtils.scala

```diff
@@ -0,0 +1,216 @@
+    // Create pipeline model to featurize the dataset
+    val oneHotEncodeCategoricals = true
+    val featuresToHashTo = FeaturizeUtilities.numFeaturesTreeOrNNBased
+    val featureColumns = dataset.columns.filter(col => col != labelColumn).toSeq
+    val featurizer = new Featurize()
+      .setFeatureColumns(Map(featuresColumn -> featureColumns))
+      .setOneHotEncodeCategoricals(oneHotEncodeCategoricals)
+      .setNumberOfFeatures(featuresToHashTo)
+    featurizer.fit(dataset)
+  }
+
+  /** Returns an integer ID for the current node.
+    * @return In cluster, returns the executor id.  In local case, returns the worker id.
+    */
+  def getId(): Int = {
+    val executorId = SparkEnv.get.executorId
+    val ctx = TaskContext.get
+    val partId = ctx.partitionId
+    // If driver, this is only in test scenario, make each partition a separate worker
+    val id = if (executorId == "driver") partId else executorId
+    val idAsInt = id.toString.toInt
+    idAsInt
+  }
+
+  /** Returns the executors.
+    * @param dataset The dataset containing the current spark session.
+    * @param defaultListenPort The default port to listen on.
+    * @return List of executors in host:port format.
+    */
+  def getExecutors(dataset: Dataset[_], defaultListenPort: Int): Array[String] = {
+    val blockManager = BlockManagerUtils.getBlockManager(dataset)
+    blockManager.master.getMemoryStatus.flatMap({ case (blockManagerId, _) =>
+      if (blockManagerId.executorId == "driver") None
+      else Some(blockManagerId.host + ":" + (defaultListenPort + blockManagerId.executorId.toInt))
+    }).toArray
+  }
+
+  /**
+    * Returns the number of executors.
+    * @param dataset The dataset containing the current spark session.
+    * @param defaultListenPort The default port to listen on.
+    * @return The number of executors.
+    */
+  def getNumExecutors(dataset: Dataset[_], defaultListenPort: Int): Int = {
+    val executors = getExecutors(dataset, defaultListenPort)
+    if (executors.isEmpty) {
+      // Case when run locally in local[*]
+      val blockManager = BlockManagerUtils.getBlockManager(dataset)
+      blockManager.master.getMemoryStatus.size
+    } else {
+      executors.length
+    }
+  }
+
+  /** Returns the executor node ips and ports.
+    * @param data The input dataframe.
+    * @param defaultListenPort The default listen port.
+    * @return List of nodes as comma separated string and count.
+    */
+  def getNodes(data: DataFrame, defaultListenPort: Int): (String, Int) = {
+    val nodes = getExecutors(data, defaultListenPort)
+    if (nodes.isEmpty) {
+      // Running in local[*]
+      getNodesFromPartitions(data, defaultListenPort)
+    } else {
+      // Running on cluster, include all workers with driver excluded
+      (nodes.sorted.distinct.reduceLeft((val1, val2) => val1 + "," + val2), nodes.size)
+    }
+  }
+
+  def getNodesFromPartitions(processedData: DataFrame, defaultListenPort: Int): (String, Int) = {
```

Comment:
I thought you did not need this anymore?

Reply:
I only need it for the local[*] case, all of the executor id's are driver when using local[*] (which seems kind of like a bug)

Reply:
For local[*], can't you just use localhost?